{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f58bc8f5",
   "metadata": {},
   "source": [
    "# GSM of C. necator: Comparisons of iCN1361 to RehMBEL1391"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d274c67e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.1.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.1.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import cobra\n",
    "#from cameo import pfba\n",
    "import json\n",
    "from itertools import chain\n",
    "from optlang.symbolics import Zero\n",
    "from cobra.flux_analysis.variability import flux_variability_analysis\n",
    "import numpy as np\n",
    "from cobra.sampling import sample\n",
    "from cobra.flux_analysis import pfba\n",
    "import matplotlib.pyplot as plt\n",
    "from cameo import models\n",
    "from memote.support.consistency import detect_energy_generating_cycles\n",
    "from bioservices import kegg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1fb70efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### load the models\n",
    "cobra.core.configuration.Configuration.solver = 'cplex'\n",
    "# load the iCN1361\n",
    "m = cobra.io.read_sbml_model('../Model/iCN1361.xml')\n",
    "# load the RehMBEL1391 - downloaded from Michael Jahn's github\n",
    "m2 = cobra.io.read_sbml_model('Data/Jahn2022_files/RehMBEL1391_sbml_L3V1.xml')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c041bbf",
   "metadata": {},
   "source": [
    "## Model properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "a9c386fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reactions:  1292 1538\n",
      "number of transport reactions:  98 384\n",
      "number of internal reactions:  1194 1154\n",
      "number of cytosol metabolites:  1263 1172\n",
      "number of genes 1362 1345\n",
      "Balanced reactions:  0.9966555183946488 0.4922850844966936\n",
      "Blocked reactions:  443 593\n",
      "Reactions active:  751 561\n",
      "Erroneous energy-generating cycles:  0 0\n",
      "Unconserved metabolites:  0 235\n",
      "MEMOTE overall score: 90% 60%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# number of reactions\n",
    "print('number of reactions: ', len(m.reactions), len(m2.reactions))\n",
    "\n",
    "# number of transporters in iCN1361\n",
    "transporter = []\n",
    "for i in m.reactions:\n",
    "    if 'EX_' in i.id:\n",
    "        transporter.append(i.id)\n",
    "    if 'transport' in i.name:\n",
    "        transporter.append(i.id)\n",
    "transporter = list(set(transporter))\n",
    "# number of transporters in RehBMEL1391\n",
    "transporter2 = []\n",
    "for i in m2.reactions:\n",
    "    if 'EX_' in i.id:\n",
    "        transporter2.append(i.id)\n",
    "    if 'transport' in i.name:\n",
    "        transporter2.append(i.id)\n",
    "transporter2 = list(set(transporter2))        \n",
    "# number of internal reactions in each model\n",
    "print('number of transport reactions: ', len(transporter), len(transporter2))\n",
    "print('number of internal reactions: ', len(m.reactions) - len(transporter), len(m2.reactions) - len(transporter2))\n",
    "\n",
    "# number of internal metabolites in iCN1361\n",
    "mets_c = []\n",
    "for i in m.metabolites:\n",
    "    if i.compartment == 'c':\n",
    "        mets_c.append(i.id)\n",
    "        \n",
    "# number of internal metabolites in RehMBEL1391        \n",
    "mets2_c = []\n",
    "for i in m2.metabolites:\n",
    "    if i.compartment == 'c':\n",
    "        mets2_c.append(i.id)\n",
    "\n",
    "print('number of cytosol metabolites: ', len(mets_c), len(mets2_c))\n",
    "\n",
    "# any genes in iCN1361 that are not associated with a reaction\n",
    "unassigned = []\n",
    "for i in m.genes:\n",
    "    if len(i.reactions) == 0:\n",
    "        unassigned.append(i.id)\n",
    "        \n",
    "\n",
    "# any genes in RehMBEL1391 that are not associated with a reaction\n",
    "unassigned2 = []\n",
    "for i in m2.genes:\n",
    "    if len(i.reactions) == 0:\n",
    "        unassigned2.append(i.id)\n",
    "        \n",
    "print('number of genes', len(m.genes)- len(unassigned), len(m2.genes)- len(unassigned2))\n",
    "\n",
    "\n",
    "# number of mass balanced reactions in iCN1361\n",
    "balanced = []\n",
    "reacs = []\n",
    "for i in m.reactions:\n",
    "    if 'EX_' not in i.id:\n",
    "        if i.check_mass_balance() == {}:\n",
    "            balanced.append(i.id)\n",
    "        reacs.append(i.id)\n",
    "    \n",
    "# number of mass balanced reactions in RehMBEL1391 - note that some metabolites are missing metabolites so could come up as not balanced due to this.        \n",
    "balanced2 = []\n",
    "reacs2 = []\n",
    "for i in m2.reactions:\n",
    "    if 'EX_' not in i.id:\n",
    "        if i.check_mass_balance() == {}:\n",
    "            balanced2.append(i.id)\n",
    "        else:\n",
    "            if 'C' not in i.check_mass_balance() and 'H' not in i.check_mass_balance() and 'O' not in i.check_mass_balance() and 'P' not in i.check_mass_balance() and 'S' not in i.check_mass_balance():\n",
    "                balanced2.append(i.id)\n",
    "        reacs2.append(i.id)\n",
    "\n",
    "print('Balanced reactions: ', len(balanced)/len(reacs), len(balanced2)/len(reacs2))\n",
    "\n",
    "# Number of dead reactions in each model (with exchange reactions open)\n",
    "dead = cobra.flux_analysis.find_blocked_reactions(m, open_exchanges=True)\n",
    "dead2 = cobra.flux_analysis.find_blocked_reactions(m2, open_exchanges=True)\n",
    "\n",
    "print('Blocked reactions: ', len(dead), len(dead2))\n",
    "print('Reactions active: ', len(m.reactions) - len(transporter) - len(dead), len(m2.reactions) - len(transporter2) - len(dead2))\n",
    "\n",
    "\n",
    "\n",
    "# Energy generation - match the ids in the iCN1361 model (cannot get to work in MEMOTE without)\n",
    "m.metabolites.M_ATP.annotation['metanetx.chemical'] = 'MNXM3'\n",
    "m.metabolites.M_GTP.annotation['metanetx.chemical'] = 'MNXM51'\n",
    "m.metabolites.M_NADH.annotation['metanetx.chemical'] = 'MNXM10'\n",
    "m.metabolites.M_NADPH.annotation['metanetx.chemical'] = 'MNXM6'\n",
    "m.metabolites.M_TTP.annotation['metanetx.chemical'] = 'MNXM394'\n",
    "m.metabolites.M_UTP.annotation['metanetx.chemical'] = 'MNXM121'\n",
    "m.metabolites.M_CTP.annotation['metanetx.chemical'] = 'MNXM63'\n",
    "m.metabolites.M_ITP.annotation['metanetx.chemical'] = 'MNXM423'\n",
    "m.metabolites.M_FADH2.annotation['metanetx.chemical'] = 'MNXM38'\n",
    "m.metabolites.M_Reduced_flavodoxins.annotation['metanetx.chemical'] = 'MNXM208'\n",
    "m.metabolites.M_Ubiquinols.annotation['metanetx.chemical'] = 'MNXM191'\n",
    "m.metabolites.M_Menaquinols.annotation['metanetx.chemical'] = 'MNXM223'\n",
    "m.metabolites.M_ACETYL_COA.annotation['metanetx.chemical'] = 'MNXM21'\n",
    "m.metabolites.M_GLT.annotation['metanetx.chemical'] = 'MNXM89557'\n",
    "m.metabolites.M_CO_A.annotation['metanetx.chemical'] = 'MNXM12'\n",
    "m.metabolites.M_2_KETOGLUTARATE.annotation['metanetx.chemical'] = 'MNXM20'\n",
    "m.metabolites.M_ADP.annotation['metanetx.chemical'] = 'MNXM7'\n",
    "m.metabolites.M_CDP.annotation['metanetx.chemical'] = 'MNXM220'\n",
    "m.metabolites.M_GDP.annotation['metanetx.chemical'] = 'MNXM30'\n",
    "m.metabolites.M_UDP.annotation['metanetx.chemical'] = 'MNXM17'\n",
    "m.metabolites.M_IDP.annotation['metanetx.chemical'] = 'MNXM495'\n",
    "m.metabolites.M_NADP.annotation['metanetx.chemical'] = 'MNXM5'\n",
    "m.metabolites.M_NAD.annotation['metanetx.chemical'] = 'MNXM8'\n",
    "m.metabolites.M_FAD.annotation['metanetx.chemical'] = 'MNXM33'\n",
    "m.metabolites.M_Oxidized_flavodoxins.annotation['metanetx.chemical'] = 'MNXM119'\n",
    "m.metabolites.M_Ubiquinones.annotation['metanetx.chemical'] = 'MNXM232'\n",
    "m.metabolites.M_PROTON.annotation['metanetx.chemical'] = 'MNXM1'\n",
    "m.metabolites.M_WATER.annotation['metanetx.chemical'] = 'MNXM2'\n",
    "m.metabolites.M_AMMONIUM.annotation['metanetx.chemical'] = 'MNXM15'\n",
    "\n",
    "\n",
    "\n",
    "model = m\n",
    "metabolite_id = 'MNXM3'\n",
    "ENERGY_COUPLES = {\n",
    "    \"MNXM3\": \"MNXM7\",\n",
    "    \"MNXM63\": \"MNXM220\",\n",
    "    \"MNXM51\": \"MNXM30\",\n",
    "    \"MNXM121\": \"MNXM17\",\n",
    "    \"MNXM423\": \"MNXM495\",\n",
    "    \"MNXM6\": \"MNXM5\",\n",
    "    \"MNXM10\": \"MNXM8\",\n",
    "    \"MNXM38\": \"MNXM33\",\n",
    "    \"MNXM191\": \"MNXM232\",\n",
    "    \"MNXM21\": \"MNXM12\",\n",
    "    \"MNXM89557\": \"MNXM20\",\n",
    "}\n",
    "\n",
    "energy_errors = []\n",
    "for i, j in ENERGY_COUPLES.items():\n",
    "    metabolite_id = i\n",
    "    energy_errors.append(detect_energy_generating_cycles(m, metabolite_id))\n",
    "    m.reactions.Dissipation.remove_from_model()\n",
    "    \n",
    "    \n",
    "energy_errors2 = []\n",
    "for i, j in ENERGY_COUPLES.items():\n",
    "    metabolite_id = i\n",
    "    energy_errors2.append(detect_energy_generating_cycles(m2, metabolite_id))\n",
    "    m2.reactions.Dissipation.remove_from_model()\n",
    "\n",
    "energy_cycles_count = 0\n",
    "for i in energy_errors:\n",
    "    if i != []:\n",
    "        energy_cycles_count += 1\n",
    "        \n",
    "energy_cycles_count2 = 0\n",
    "for i in energy_errors2:\n",
    "    if i != []:\n",
    "        energy_cycles_count2 += 1\n",
    "        \n",
    "\n",
    "print('Erroneous energy-generating cycles: ', energy_cycles_count, energy_cycles_count2)\n",
    "print('Unconserved metabolites: ', 0, 235) # taken from the MEMOTE report\n",
    "print('MEMOTE overall score: ' '90%', '60%') # taken from the MEMOTE report \n",
    "print('')\n",
    "\n",
    "# notes:\n",
    "# Only 162 of the 207 transporters are active in the current model (i.e. not in the set of dead reactions)\n",
    "# and only 190 of the transporters are missing a gene association, so it's not clear how the original model derived these transporters...\n",
    "# We include transporters only for those that we know can be transported into the model, but since we're only interested in growth on fructose any other transporters would be dead in the simulations presented anyway. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d987d392",
   "metadata": {},
   "source": [
    "### Notable differences between the model constructions\n",
    "\n",
    "•\tTransport reactions: in our model we added transporters only for minimal media substrates and for carbon sources known to enable growth. Annotating transport reactions is much more challenging, so we only added those for which experimental evidence exists to show the bacteria is capable of utilising the transporter. Transporters were also added for metabolites commonly produced by bacteria, in the same way as [ref - BioCyc]. The transporters in RehMBEL1361 were defined in the original model, but the majority lack a gene annotation so it is unclear whether the transporter is definitely present. \n",
    "\n",
    "•\tCell wall metabolism: in our model we include the full pathway for each fatty acid included in the model, where RehMBEL1391 uses a lumped version. Using the full pathway can be useful when integrating transcriptomics data and/or proteomics data (otherwise the genes will need to be weighted according to the individual reactions involved the lumped version), whereas a lumped version can be beneficial when reducing the network size when applying approaches such as optKnock or optGene.  (fatty acids and LPS)\n",
    "\n",
    "•\tBiomass equation: we have used the same biomass as defined in the original RehMBEL1391 model and included the correction identified in [1]. In [2], the original model manuscript, they reference the cellular composition being taken from E. coli values for the protein content, yet the value is considerably higher. We therefore have modified the composition to be close to the E. coli iJO1366 model (see SupplementaryFile1_GSMInfo – Table S4). The way in which the biomass reaction is defined is slightly different in the two models. In iCN1361 we define one biomass reaction that includes all individual biomass components (i.e. alanine + aspartate + dATP + dGTP +  Biomass ), whereas in RehBMEL1391 they include separate reactions for each macromolecule (i.e. protein + DNA + …  Biomass). However, these are just different ways of formulating the biomass equation and do not affect the model results.\n",
    "\n",
    "•\tEnzyme specificity: iCN1361 was constructed using the BioCyc database, whereas the RehMBEL1391 has assigned gene annotations using KEGG and UniProt. These databases sometimes differ in the number of reactions assigned to a gene, due to differences in the predicted substrate specificity of the enzyme. \n",
    "\n",
    "\n",
    "\n",
    "$\\bf{References}\\\\$\n",
    "Weaver, Daniel S., et al. \"A genome-scale metabolic flux model of $\\it{Escherichia~coli}$ K-12 derived from the EcoCyc database\", BMC Systems Biology (2014)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d845e0f",
   "metadata": {},
   "source": [
    "## Model validation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c553bee9",
   "metadata": {},
   "source": [
    "### Carbon source utilisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1133d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "carbon_sources = pd.read_excel('Data/Carbon_source_utilisation_phenotypes.xlsx', sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "134fdd17",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "substrates = list(carbon_sources['Metabolite ID'])\n",
    "substrates2 = []\n",
    "match = {}\n",
    "match2 = {}\n",
    "for i in substrates:\n",
    "    if i[0:3] != 'M_':\n",
    "        substrates2.append('M_'+i)\n",
    "        match['M_' + i] = i\n",
    "        match2[i] = 'M_' + i\n",
    "    else:\n",
    "        substrates2.append(i)\n",
    "        match[i] = i\n",
    "        match2[i] = i\n",
    "        \n",
    "\n",
    "# map the exchange reactions in RehMBEL1391 to the tested carbon sources\n",
    "reh_exc = {'M_2_PG' : 'EX_2pg_e',\n",
    "           'M_G3P' : 'EX_3pg_e',\n",
    "           'M_ACET' : 'EX_ac_e',\n",
    "           'M_CIT' : 'EX_cit_e',\n",
    "           'M_BETA_D_FRUCTOSE' : 'EX_fru_e',\n",
    "           'M_L_LACTATE' : 'EX_lac__L_e',\n",
    "           'M_D_LACTATE' : 'EX_lac__D_e',\n",
    "           'M_FUM' : 'EX_fum_e',\n",
    "           'M_GLYOX' : 'EX_glx_e',\n",
    "           'M_THREO_DS_ISO_CITRATE' : 'EX_icit_e',\n",
    "           'M_L_ALPHA_ALANINE' : 'EX_ala__L_e',\n",
    "           'M_ASN' : 'EX_asn__L_e',\n",
    "           'M_L_ASPARTATE' : 'EX_asp__L_e',\n",
    "           'M_GLT' : 'EX_glu__L_e',\n",
    "           'M_HIS' : 'EX_his__L_e',\n",
    "           'M_LEU' : 'EX_leu__L_e',\n",
    "           'M_PHE' : 'EX_phe__L_e',\n",
    "           'M_PRO' : 'EX_pro__L_e',\n",
    "           'M_SER' : 'EX_ser__L_e',\n",
    "           'M_THR' : 'EX_thr__L_e',\n",
    "           'M_TRP' : 'EX_trp__L_e',\n",
    "           'M_NIACINE' : 'EX_nac_e',\n",
    "           'M_PHENOL' : 'EX_phenol_e',\n",
    "           'M_PHENYLACETATE' : 'EX_pac_e',\n",
    "           'M_PROPIONATE' : 'EX_ppa_e',\n",
    "           'M_PYRUVATE' : 'EX_pyr_e',\n",
    "           'M_SUC' : 'EX_succ_e',\n",
    "           'M_TREHALOSE' : 'EX_tre_e',\n",
    "           'M_CARNITINE' : 'EX_crn_e',\n",
    "           'M_GLC_6_P' : 'EX_g6p_e',\n",
    "           'M_INOSINE' : 'EX_ins_e',\n",
    "           'M_PUTRESCINE' : 'EX_ptrc_e',\n",
    "           'M_THYMIDINE' : 'EX_thymd_e',\n",
    "           'M_URIDINE' : 'EX_uri_e',\n",
    "           'M_ETOH' : 'EX_eth_e',\n",
    "           'M_MAL' : 'EX_mal_e',\n",
    "           'M_4_AMINO_BUTYRATE' : 'EX_gaba_e',\n",
    "           'M_GLYCEROL' : 'EX_gl_e',\n",
    "           'M_GLUCONATE' : 'EX_gluc_e',\n",
    "           'M_D_ALANINE' : 'EX_dala_e',\n",
    "           'M_CIS_ACONITATE' : 'EX_acon__C_e',\n",
    "           'M_BENZALDEHYDE' : 'EX_bzald_e',\n",
    "           'M_BENZOATE' : 'EX_benzot_e',\n",
    "           'M_ANTHRANILATE' : 'EX_an_e',\n",
    "           'M_2_OXOBUTANOATE' : 'EX_obut_e',\n",
    "           'M_4_hydroxybenzoate' : 'EX_4hb_e',\n",
    "           'M_CPD_335' : 'EX_r3hbn_e',\n",
    "           'M_FORMATE' : 'EX_formate_e',\n",
    "           'M_ADIPATE' : 'EX_adip_e',\n",
    "           'M_CATECHOL' : 'EX_catech_e',\n",
    "           'M_BUTYRIC_ACID' : 'EX_butn_e',\n",
    "           'M_3_HYDROXYBENZOATE' : 'EX_3hbenzot_e', \n",
    "           'M_4_FLUOROBENZOATE' : 'EX_4flrbz_e',\n",
    "           'M_6_CHQ' : 'EX_6chq_e',\n",
    "           'M_GLYCOLLATE' : 'EX_glycolate_e',\n",
    "           'M_KYNURENATE' : 'EX_knt_e',\n",
    "           'M_CPD_108' : 'EX_4cresol_e',    \n",
    "           'M_2_HYDROXYBUTANOIC_ACID' : 'EX_2hba_e',\n",
    "           'M_PHENYLGLYOXYLATE' : 'EX_aobzac_e',\n",
    "           'M_BIPHENYL_23_DIOL' : 'EX_bp23d_e',\n",
    "           'M_CIS_CIS_MUCONATE' : 'EX_ccmuc_e',\n",
    "           'M_4_HYDROXY_BUTYRATE' : 'EX_4hbt_e',\n",
    "           'M_CPD_633' : 'EX_gensa_e',\n",
    "           'M_CPD_14736' : 'EX_kn_e',\n",
    "           'M_CPD_741' : 'EX_mclact_e',\n",
    "           'M_CPD_7836' : 'EX_c140_e',\n",
    "           'M_OLEATE_CPD' : 'EX_c181_e',\n",
    "           'M_N_ACETYL_D_GLUCOSAMINE_e' : 'EX_naga_e',\n",
    "           'M_PALMITATE' : 'EX_c160_e',\n",
    "           'M_4_HYDROXYPHENYLACETATE' : 'EX_4hpheac_e',\n",
    "           'M_3_4_DIHYDROXYBENZOATE' : 'EX_34dhb_e',\n",
    "           'M_UROCANATE' : 'EX_urocan_e',\n",
    "           'M_alpha-d-lactose' : 'EX_lactose_e',\n",
    "           'M_D_GALACTONATE' : 'EX_galctn__D_e',\n",
    "           'M_GLUCURONATE' : 'EX_dgluc_e',\n",
    "           'M_MANNITOL' : 'EX_mnt_e',\n",
    "           'M_MANNOSE' : 'EX_man_e',\n",
    "           'M_MELIBIOSE' : 'EX_meli_e',\n",
    "           'M_D-SERINE' : 'EX_dser_e', \n",
    "           'M_SORBITOL' : 'EX_sot_e',\n",
    "           'M_ARABINOSE' : 'EX_larabinose_e',\n",
    "           'M_L-fucose' : 'EX_fuc_e',\n",
    "           'M_L_ORNITHINE' : 'EX_orn_e',\n",
    "           'M_D-Rhamnose' : 'EX_rmn_e',\n",
    "           'M_MALTOSE' : 'EX_mlt_e',\n",
    "           'M_RIBOSE' : 'EX_rib_e',\n",
    "           'M_ETHANOL_AMINE' : 'EX_etha_c', \n",
    "           'M_ACETOIN' : 'EX_acetoin_c',\n",
    "           'M_GLC' : 'EX_glc__D_c',\n",
    "           'M_2_KETOGLUTARATE' : 'EX_akg_c',\n",
    "           'M_1,3-butanediol' : 'EX_13but_c',\n",
    "           'M_5_DEHYDROGLUCONATE' : 'EX_5dh4dglc_c',\n",
    "           'M_adonitol' : 'EX_adon_c',\n",
    "           'M_alpha-cyclodextrin' : 'EX_alpha-cyclodextrin_c',\n",
    "           'M_b-methyl-d-glucoside' : 'EX_b-methyl-d-glucoside_c',\n",
    "           'M_cellobiose' : 'EX_cellobiose_c',\n",
    "           'M_d-arabitol' : 'EX_d_arabitol_c',\n",
    "           'M_dextrin' : 'EX_dextrin_c',\n",
    "           'M_CPD_3617' : 'EX_dodec_c',\n",
    "           'M_D-galactonic acid lactone' : 'EX_dgalactonic_lactone_c',\n",
    "           'M_ALPHA_D_GALACTOSE' : 'EX_gal_c',\n",
    "           'M_D_GALACTURONATE' : 'EX_galur_c',\n",
    "           'M_GLUCOSAMINATE' : 'EX_glucosaminate_c',\n",
    "           'M_Glycerol-1-phosphate' : 'EX_glyc1p_c',\n",
    "           'M_CPD-1099' : 'EX_raff_c',\n",
    "           'M_D-GLUCARATE' : 'EX_glcr_c',\n",
    "           'M_Gentiobiose' : 'EX_gentibiose_c',\n",
    "           'M_GLC_1_P' : 'EX_g1p_c',\n",
    "           'M_Glycogen' : 'EX_glycogen_c',\n",
    "           'M_glycyl-l-aspartic acid' : 'EX_glycl_asp_c',\n",
    "           'M_hydroxy-l-proline' : 'EX_4hpro_c',\n",
    "           'M_ERYTHRITOL' : 'EX_erythritol_c',\n",
    "           'M_Isomaltose' : 'EX_imal_c',\n",
    "           'M_ITACONATE' : 'EX_itcn_c',\n",
    "           'M_CPD-3561' : 'EX_lactulose_c', \n",
    "           'M_L-Mandelate' : 'EX_mand_c',\n",
    "           'M_MALONATE' : 'EX_malonate_c',\n",
    "           'M_CPD-112' : 'EX_3cresol_c',  \n",
    "           'M_Inositols' : 'EX_inositols_c',\n",
    "           'M_N_acetyl_D_galactosamine' : 'EX_ndgal_c',\n",
    "           'M_CPD-109' : 'EX_ocresol_c',\n",
    "           'M_PHENYLETHYLAMINE' : 'EX_peamn_c',\n",
    "           'M_QUINATE' : 'EX_quin_c',\n",
    "           'M_CPD-1142' : 'EX_salicin_c',\n",
    "           'M_SHIKIMATE' : 'EX_skm_c',\n",
    "           'M_SUCROSE' : 'EX_sucrose_c',\n",
    "           'M_CPD-13399' : 'EX_tura_c',\n",
    "           'M_XYLITOL' : 'EX_xylt_c',\n",
    "           'M_XYLOSE' : 'EX_xyl__D_c',\n",
    "           'M_Alpha-lactose' : 'EX_lactose_c'\n",
    "            }\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "03cd1359",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test without adding transporters for those missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "39555267",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tested carbon sources:  131\n",
      "Number of true positives:  60\n",
      "Number of false positives:  4\n",
      "Number of true negatives:  63\n",
      "Number of false negatives:  4\n",
      "Overall accuracy:  0.9389312977099237\n"
     ]
    }
   ],
   "source": [
    "m2 = cobra.io.read_sbml_model('Data/Jahn2022_files/RehMBEL1391_sbml_L3V1.xml')\n",
    "\n",
    "\n",
    "# set the constraints\n",
    "m2.reactions.EX_fru_e.bounds = (0.0, 0.0)\n",
    "m2.reactions.Maintenance.bounds = (3.0, 1000.0)\n",
    "m2.objective = 'Biomass'\n",
    "m2.objective_direction = 'max'\n",
    "m2.reactions.Biomass.bounds = (0.0, 1000.0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "growth_phenotypes = {}\n",
    "# check to see if metabolite in the model\n",
    "for met, exc in reh_exc.items():\n",
    "    #print(met)\n",
    "    \n",
    "    # check if in the model\n",
    "    run = 1\n",
    "    with m2:\n",
    "        if exc in m2.reactions:\n",
    "            m2.reactions.get_by_id(exc).bounds = (-10.0, 1000.0)\n",
    "        \n",
    "        \n",
    "        \n",
    "        elif exc not in m2.reactions:\n",
    "            met_id = exc.split('EX_')[1]\n",
    "            \n",
    "            if met_id not in m2.metabolites:\n",
    "                growth_phenotypes[met] = 0.0\n",
    "                run = 0\n",
    "                \n",
    "            else:\n",
    "                growth_phenotypes[met] = 0.0\n",
    "                run = 0\n",
    "                \n",
    "\n",
    "                \n",
    "                \n",
    "        # check whether solution exists\n",
    "        if run == 1:\n",
    "            try:                \n",
    "                sol = pfba(m2)\n",
    "                #print(reaction.id, sol[reaction.id], sol['R_Biomass'])\n",
    "                if sol['Biomass'] > 0.01:\n",
    "                    growth_phenotypes[met] = 1.0\n",
    "                else:\n",
    "                    growth_phenotypes[met] = 0.0\n",
    "            except:\n",
    "                growth_phenotypes[met] = 0.0\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "### Print results to dataframe\n",
    "\n",
    "\n",
    "FP = 0\n",
    "FN = 0\n",
    "TP = 0\n",
    "TN = 0\n",
    "false_positives = []\n",
    "false_negatives = []\n",
    "\n",
    "for i, j in growth_phenotypes.items():\n",
    "    if j == 1.0:\n",
    "        for ind, met in enumerate(carbon_sources['Metabolite ID']):\n",
    "            if met == match[i]:\n",
    "                if carbon_sources['Experimental'][ind] == 'X':\n",
    "                    FP += 1\n",
    "                    false_positives.append(met)\n",
    "                    \n",
    "                elif carbon_sources['Experimental'][ind] == 'O':\n",
    "                    TP += 1\n",
    "                    \n",
    "                    \n",
    "    if j == 0.0:\n",
    "        for ind, met in enumerate(carbon_sources['Metabolite ID']):\n",
    "            if met == match[i]:\n",
    "                if carbon_sources['Experimental'][ind] == 'X':\n",
    "                    TN += 1\n",
    "                    \n",
    "                    \n",
    "                elif carbon_sources['Experimental'][ind] == 'O':\n",
    "                    FN += 1\n",
    "                    false_negatives.append(met)\n",
    "\n",
    "\n",
    "print('Number of tested carbon sources: ', len(substrates2))\n",
    "print('Number of true positives: ', TP)\n",
    "print('Number of false positives: ', FP)\n",
    "print('Number of true negatives: ', TN)\n",
    "print('Number of false negatives: ', FN)\n",
    "overall_accuracy = (TP+TN)/(TP+TN+FP+FN)\n",
    "print('Overall accuracy: ',  overall_accuracy)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d233106",
   "metadata": {},
   "source": [
    "### Test whilst adding transporters for those missing - for a better comparison to iCN1361 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "58c6eeae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tested carbon sources:  131\n",
      "Number of true positives:  60\n",
      "Number of false positives:  13\n",
      "Number of true negatives:  54\n",
      "Number of false negatives:  4\n",
      "Overall accuracy:  0.8702290076335878\n"
     ]
    }
   ],
   "source": [
    "m2 = cobra.io.read_sbml_model('Data/Jahn2022_files/RehMBEL1391_sbml_L3V1.xml')\n",
    "from cobra.core import Reaction\n",
    "\n",
    "# set the constraints\n",
    "m2.reactions.EX_fru_e.bounds = (0.0, 0.0)\n",
    "m2.reactions.Maintenance.bounds = (3.0, 1000.0)\n",
    "m2.objective = 'Biomass'\n",
    "m2.objective_direction = 'max'\n",
    "m2.reactions.Biomass.bounds = (0.0, 1000.0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "growth_phenotypes = {}\n",
    "# check to see if metabolite in the model\n",
    "for met, exc in reh_exc.items():\n",
    "    #print(met)\n",
    "    \n",
    "    # check if in the model\n",
    "    run = 1\n",
    "    with m2:\n",
    "        if exc in m2.reactions:\n",
    "            m2.reactions.get_by_id(exc).bounds = (-10.0, 1000.0)\n",
    "        \n",
    "        \n",
    "        \n",
    "        elif exc not in m2.reactions:\n",
    "            met_id = exc.split('EX_')[1]\n",
    "            \n",
    "            if met_id not in m2.metabolites:\n",
    "                growth_phenotypes[met] = 0.0\n",
    "                run = 0\n",
    "                \n",
    "            else:\n",
    "                reh_exc_new = Reaction(exc)\n",
    "                reh_exc_new.add_metabolites({m2.metabolites.get_by_id(met_id) : -1})\n",
    "                m2.add_reaction(reh_exc_new)\n",
    "                m2.reactions.get_by_id(exc).bounds = (-10.0, 0.0)\n",
    "                \n",
    "                \n",
    "                \n",
    "\n",
    "                \n",
    "                \n",
    "        # check whether solution exists\n",
    "        if run == 1:\n",
    "            try:                \n",
    "                sol = pfba(m2)\n",
    "                #print(reaction.id, sol[reaction.id], sol['R_Biomass'])\n",
    "                if sol['Biomass'] > 0.01:\n",
    "                    growth_phenotypes[met] = 1.0\n",
    "                else:\n",
    "                    growth_phenotypes[met] = 0.0\n",
    "            except:\n",
    "                growth_phenotypes[met] = 0.0\n",
    "        \n",
    "        \n",
    "        \n",
    "### Print results to dataframe\n",
    "\n",
    "\n",
    "FP = 0\n",
    "FN = 0\n",
    "TP = 0\n",
    "TN = 0\n",
    "false_positives = []\n",
    "false_negatives = []\n",
    "\n",
    "for i, j in growth_phenotypes.items():\n",
    "    if j == 1.0:\n",
    "        for ind, met in enumerate(carbon_sources['Metabolite ID']):\n",
    "            if met == match[i]:\n",
    "                if carbon_sources['Experimental'][ind] == 'X':\n",
    "                    FP += 1\n",
    "                    false_positives.append(met)\n",
    "                    \n",
    "                elif carbon_sources['Experimental'][ind] == 'O':\n",
    "                    TP += 1\n",
    "                    \n",
    "                    \n",
    "    if j == 0.0:\n",
    "        for ind, met in enumerate(carbon_sources['Metabolite ID']):\n",
    "            if met == match[i]:\n",
    "                if carbon_sources['Experimental'][ind] == 'X':\n",
    "                    TN += 1\n",
    "                    \n",
    "                    \n",
    "                elif carbon_sources['Experimental'][ind] == 'O':\n",
    "                    FN += 1\n",
    "                    false_negatives.append(met)\n",
    "\n",
    "\n",
    "print('Number of tested carbon sources: ', len(substrates2))\n",
    "print('Number of true positives: ', TP)\n",
    "print('Number of false positives: ', FP)\n",
    "print('Number of true negatives: ', TN)\n",
    "print('Number of false negatives: ', FN)\n",
    "overall_accuracy = (TP+TN)/(TP+TN+FP+FN)\n",
    "print('Overall accuracy: ',  overall_accuracy)        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5aea545",
   "metadata": {},
   "source": [
    "### Comparison to iCN1361\n",
    "9/15 false positives are overlapping with the false positives found in iCN1361. If we were assume a transporter is not available for these metabolites, as is the case in the RehMBEL1391 model, then the number of false positives would decrease to 6 and the overall accuracy increase to 93%.\n",
    "\n",
    "\n",
    "|  | iCN1361 | RehMBEL1391 |\n",
    "|--|---------|-------------|\n",
    "| Number of TPs | 62 | 60 |\n",
    "| Number of FPs | 6 | 4 |\n",
    "| Number of TNs | 61 | 63 |\n",
    "| Number of FNs | 3 | 4 |\n",
    "| Overall accuracy | 93% | 94% |\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7469b539",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02a58b15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3217a963",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f601be04",
   "metadata": {},
   "source": [
    "### Flux predictions - growth on fructose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "63fd51c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Growth rate:  0.233978640837583\n",
      "EX_fe2_e -5.194325826594343e-05\n",
      "EX_BIOMASS_c 0.233978640837583\n",
      "EX_pi_e -6.5503466997821445\n",
      "EX_so4_e -0.04754679960460524\n",
      "EX_fru_e -2.6\n",
      "EX_nh4_e -2.491430188299773\n",
      "EX_co2_e 5.882062496419901\n",
      "EX_o2_e -5.539665630907597\n",
      "EX_h2o_e 20.411560733340526\n",
      "EX_h_e -8.795573784200476\n"
     ]
    }
   ],
   "source": [
    "m2 = cobra.io.read_sbml_model('Data/Jahn2022_files/RehMBEL1391_sbml_L3V1.xml')\n",
    "\n",
    "m2.reactions.EX_fru_e.bounds = (-2.6, -2.6)\n",
    "m2.reactions.FRUpts.bounds = (0.0, 0.0) # previous study found the ABC is the active reaction [ref]\n",
    "m2.reactions.FRUpts2.bounds = (0.0, 0.0)\n",
    "#m2.reactions.EX_h_e.bounds = (0.0, 0.0)\n",
    "m2.reactions.Maintenance.bounds = (3.0, 1000.0) # wasn't already constrained - do they include somewhere else?\n",
    "m2.objective = 'Biomass'\n",
    "sol_fru2 = pfba(m2)\n",
    "\n",
    "print('Growth rate: ', sol_fru2['Biomass'])\n",
    "for i in m2.reactions:\n",
    "    if 'EX_' in i.id:\n",
    "        if abs(sol_fru2[i.id]) > 0.0:\n",
    "            print(i.id, sol_fru2[i.id])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd00c20",
   "metadata": {},
   "source": [
    "\n",
    "| Flux             | Experimental  | iCN1361 | RehMBEL1391. | iJO1366 |\n",
    "|------------------|---------------|---------|-------------|---------|\n",
    "| Growth rate      | 0.25 +/- 0.08 | 0.25    | 0.23       | 0.25    |\n",
    "| Fructose uptake  | 2.6 +/- 0.9   | 2.6*    | 2.6*        | 2.6*    |\n",
    "| Oxygen uptake    | 4.6 +/- 0.3   | 4.91     | 5.54         | 5.0     |\n",
    "| Ammonium uptake  |               | 2.63     | 2.49         | 2.65     |\n",
    "| Sulfate uptake   |               | 0.04    | 0.05        | 0.06    |\n",
    "| Phosphate uptake |               | 0.22    | 6.55       | 0.24    |\n",
    "| CO2 production   |               | 5.21    | 5.54         | 5.5     |\n",
    "| H2O production   |               | 11.95    | 20.41        | 12.0   |\n",
    "| H+ production    |               | 2.31    | -8.80       | 2.26   |\n",
    "\n",
    "In general, iCN1361 and RehMBEL1391 show very similar uptake/production flux and growth rate for growth on fructose. The growth rate is slightly lower in RehMBEL1391 and oxygen uptake slightly higher when compared to experimental results. Also, notable is that RehMBEL1391 consumes protons rather than produces protons, which is probably a result of some reactions not being balanced. If we don't allow protons being transported inside the cell then the growth rate reduces to only 0.2 1/h. The phosphate uptake is also surprisingly high compared to the E. coli model. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebde3d6a",
   "metadata": {},
   "source": [
    "### 13-C comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "be29e0a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "c13_model = pd.read_excel('Data/C13_MFA_results.xlsx', sheet_name='Sheet1')\n",
    "reac_mappings = pd.read_excel('Data/ReactionName_mappings.xlsx', sheet_name='Sheet1')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f65472",
   "metadata": {},
   "source": [
    "#### Map the central carbon metabolism reactions in RehMBEL1391 to iCN1361"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ece7485c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "reac_map = {}\n",
    "reac_map['R_PGLUCISOM_RXN'] = 'PGI'\n",
    "reac_map['R_FRUCTOKINASE_RXN'] = 'HEXf'\n",
    "reac_map['R_GLU6PDEHYDROG_RXN'] = 'G6PDH2r'\n",
    "reac_map['R_6PGLUCONOLACT_RXN'] = 'PGL'\n",
    "reac_map['R_PGLUCONDEHYDRAT_RXN'] = 'EDD'\n",
    "reac_map['R_KDPGALDOL_RXN'] = 'EDA'\n",
    "reac_map['R_2TRANSKETO_RXN'] = 'TKT2'\n",
    "reac_map['R_TRANSALDOL_RXN'] = 'TAh'\n",
    "reac_map['R_1TRANSKETO_RXN'] = 'TKT1'\n",
    "reac_map['R_SEDOHEPTULOSE_BISPHOSPHATASE_RXN'] = 'missing'\n",
    "reac_map['R_SEDOBISALDOL_RXN'] = 'missing'\n",
    "reac_map['R_RIB5PISOM_RXN'] = 'RPI'\n",
    "reac_map['R_RIBULP3EPIM_RXN'] = 'RPE'\n",
    "reac_map['R_PHOSPHORIBULOKINASE_RXN'] = 'PRUK'\n",
    "reac_map['R_RIBULOSE_BISPHOSPHATE_CARBOXYLASE_RXN'] = 'RBPC'\n",
    "reac_map['R_PHOSGLYPHOS_RXN'] = 'PGK'\n",
    "reac_map['R_3PGAREARR_RXN'] = 'PGM'\n",
    "reac_map['R_GAPOXNPHOSPHN_RXN'] = 'GAPD'\n",
    "reac_map['R_F16ALDOLASE_RXN'] = 'FBA'\n",
    "reac_map['R_TRIOSEPISOMERIZATION_RXN'] = 'TPI'\n",
    "reac_map['R_F16BDEPHOS_RXN'] = 'FBP'\n",
    "reac_map['R_2PGADEHYDRAT_RXN'] = 'ENO'\n",
    "reac_map['R_PEPSYNTH_RXN'] = 'PPS'\n",
    "reac_map['R_PEPDEPHOS_RXN'] = 'PYK'\n",
    "reac_map['R_PEPCARBOX_RXN'] = 'PPC'\n",
    "reac_map['R_PEPCARBOX_RXN2'] = 'PPC'\n",
    "reac_map['R_MALIC_NADP_RXN'] = 'ME2'\n",
    "reac_map['R_MALATE_DEH_RXN'] = 'ME1'\n",
    "reac_map['R_RXN0_1134'] = 'PDH1'\n",
    "reac_map['R_OXALODECARB_RXN'] = 'missing'\n",
    "reac_map['R_CITSYN_RXN'] = 'CS'\n",
    "reac_map['R_ACONITATEDEHYDR_RXN'] = 'ACONT1'\n",
    "reac_map['R_ISOCIT_CLEAV_RXN'] = 'ICL'\n",
    "reac_map['R_ISOCITDEH_RXN'] = 'ICDHyr'\n",
    "reac_map['R_ISOCITRATE_DEHYDROGENASE_NAD_RXN'] = 'ICDHx'\n",
    "reac_map['R_2OXOGLUTDECARB_RXN'] = 'AKGDH'\n",
    "reac_map['R_SUCCCOASYN_RXN'] = 'SUCOAS'\n",
    "reac_map['R_SUCCINATE_DEHYDROGENASE_UBIQUINONE_RXN'] = 'SUCDi'\n",
    "reac_map['R_FUMHYDR_RXN'] = 'FUM'\n",
    "reac_map['R_MALSYN_RXN'] = 'MALS'\n",
    "reac_map['R_MALATE_DEH_RXN'] = 'MDH'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ff475b99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify the reactions that are in the reverse direction\n",
    "change_dir = ['TKT1', 'RBPC', 'TPI', 'PYK', 'PPC', 'FUM']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41cdea4a",
   "metadata": {},
   "source": [
    "#### Calculate the correlation to the 13-C data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "04878257",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R correlation coefficient:  0.9265077646334658 p-value:  1.0579177759124846e-14\n",
      "R-Squared correlation coefficient:  0.8584166379261018\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAxkklEQVR4nO3dd3yV9fn/8dcVVtgzgOwpiAQXIO6FCydYq7bWUVuq1I6fFcGJo06so62gVKvVr1VbGYIbqijWYsFBwoYgK+xNgADJuX5/3Cc1hpN9VpL38/E4j3Pu+/7kPlfunOTK5/4sc3dERESSTUqiAxAREYlECUpERJKSEpSIiCQlJSgREUlK1TpBmZmbmXqBiIhUQbUTHUCcKEmJiCQvi7SzWtegRESk6lKCEhGRpKQEJSIiSUkJSkREkpISlIiIJCUlKBERSUpKUCIikpSSIkGZWUcz+9jMFprZAjP7TYQyZmZ/NLPlZpZhZscmIlYREYmPZBmomwf8zt2/MrPGwJdmNt3dFxYqcz7QM/w4HhgffhYRkQQJhZyUlIjjbCstKWpQ7r7e3b8Kv94NLALaFyl2CfCyB2YDzczssDiHKiIiQO7BfMZ+sJjrX5pDrNYVTIoEVZiZdQGOAb4ocqg9sKbQ9loOTWIF5xhuZnNjEqCISA33xYqtDHl6Fs98nEWrRvXYnxeKyfskyy0+AMysETAR+K2776roedx9AjBBE8WKiETPrtyDPPLeYv7+xWo6NK/PKzcM5JSeaTF7v6RJUGZWhyA5verukyIUyQY6FtruEN4nIiIx9uGCDdz91nw2797Pz07uyi3nHE6DurFNIUmRoMzMgBeARe7+RDHFpgI3m9nrBJ0jdrr7+njFKCJSE23ancu9UxfwbuYGerdtzISf9Oeojs3i8t4Wq8atcgVhdjIwC8gECm5m3gF0AnD3Z8NJ7M/AecBe4Hp3L7GdqeAWXzJ8jyIiVYm788+5a/n9OwvJzQvxm7N6MvzUbtSpFZOuCxG7ASZFgooVJSgRkfJbuWUPd0zO5POsrQzs0oKHL0une1qjWL6lEpSIiBQvLz/EC599yxPTl1K3Vgqjh/TmqgGdYjbOqZCIb5AUbVAiIpJY87N3MnpSBvOzd3F2nzY8cElf2jZNTWhMSlAiIjVY7sF8npqxjL/MWkHzBnUZ9+NjOb9vW4Jm/8RSghIRqaE+z9rCHZMyWbl1Lz/s34E7hhxBswZ1Ex3W/yhBiYjUMDv3HeThdxfx+pw1dGrRgFd/djwn9WiV6LAOoQQlIlKDvD9/PXe/tYCtOfv5xand+O3gw6lft1aiw4pICUpEpAbYuCuXe96azwcLNtLnsCa8eN0A+rZvmuiwSqRu5iIi1Vgo5Lwxdw0PvbuIA3khfjv4cH52StdYDbitKHUzFxGpSVZszuH2SZl88e02BnVrwcPD+tG1VcNEh1VmSlAiItXMwfwQf5m1gqdmLKNe7RQeGZbOFQM6JkXX8fJQghIRqUYy1+5k1MQMFq7fxfl923LfxUfSukliB9xWlBKUiEg1sO9APk/OWMrzs1bQqlE9nr36OM7r2zbRYVWKEpSISBX32bIt3DE5k9Xb9nLVwI6MPv8Imtavk+iwKk0JSkSkitqx9wAPvrOIf365lq6tGvLazwdxQveWiQ4rapSgRESqGHfnncz13Dt1Adv3HmTE6d359Vk9Sa2TnANuK0oJSkSkClm/cx93T1nAjEUbSW/flL/9dCBHtkvuAbcVpYG6IiJVQCjkvPrf1Tz63mLyQiFuOftwfnpSV2on14Dbikrugbpm9lfgQmCTu/eNcPx04C3g2/CuSe5+f9wCFBFJkOWbcrh9UgZzVm7npB4teWhoOp1bVp0BtxWVNAkKeAn4M/ByCWVmufuF8QlHRCSxDuSFeO6TLP700XLq163FYz/ox+XHdahyA24rKmkSlLt/amZdEh2HiEgy+GbNDkZPzGDxht1c0O8wxlzUh9aNq+aA24pKmgRVRieY2TxgHXCruy+IVMjMhgPD4xqZiEgU7D2Qxx8+XMqL//6WtMb1+Ms1/Tm7T5tEh5UQSdVJIlyDeruYNqgmQMjdc8xsCPC0u/cs5XzqJCEiVcYnSzdz5+RM1m7fx9WDOnHbeb1pklr1B9yWQXJ3kiiNu+8q9PpdMxtnZq3cfUsi4xIRqaztew7wwNsLmfR1Nt3SGvKPX5zAwK4tEh1WwlWZBGVmbYGN7u5mNhBIAbYmOCwRkQpzd6bOW8f90xayc99BfnVmD355Ro9qN+C2opImQZnZa8DpQCszWwuMAeoAuPuzwA+Am8wsD9gHXOm6dyciVVT2jn3cPWU+Hy3exFEdm/HqZen0btsk0WEllaRqg4o2tUGJSLIJhZxXZq/isfcXE3K49dxeXHdiF2ql1Iyu48Wo2m1QIiJV3bKNuxk1MYOvVu/glJ6teGhoOh1bNEh0WElLCUpEJMYO5IUYN3M5z3y8nIb1avPED49i6DHta8yA24pSghIRiaEvV23n9kkZLN2Yw8VHteOei/rQqlG9RIdVJShBiYjEQM7+PB7/YAl/+89KDmuSyl+v68+ZvWvmgNuKUoISEYmyj5ds4q7J81m3cx/XDOrMyPN606ie/tyWl66YiEiUbM3Zz/1vL+Stb9bRo3Uj3rzxBI7rrAG3FaUEJSJSSe7OlG+yuX/aQnL25/Gbs3oy4ozu1KutAbeVoQQlIlIJa7fv5c7J8/lk6WaO6dSMRy/rx+FtGic6rGpBA3VFRCogP+T87fOVPP7hEgBuO7cXPzmhxg+4rSgN1BURiYYlG4IBt9+s2cHpvdJ4cGg67ZvVT3RY1Y4SlIhIGe3Py+eZj5YzbmYWTerX4ekrj+bio9ppwG2MKEGJiJTB3JXbGDUxg6zNexh6THvuvrAPLRrWTXRY1ZoSlIhICXbnHuSx95fwyuxVtG9Wn5euH8DpvVonOqwaQQlKRKQY/1q0kbumzGfDrlyuP6kLt57Ti4YacBs3utIiIkVs3r2f+6Yt4O2M9fRq05hxPz6WYzo1T3RYNY66mYuIhLk7E7/K5oG3F7LvQD43n9mDG0/rTt3aKYkOrbpTN3MRkeKs2baXOyZnMmvZFvp3bs4jl6XTo7UG3CaSalAiUq1kZMApp8CuXd/ta9IEZs2Cbt1g7FgYNw62boWWLeHGm0K0OWkVz3y2hFopxqjze/PjgZ1I0YDbeIp4sZMmQZnZX4ELgU3u3jfCcQOeBoYAe4Hr3P2rUs6pBCVSg2RkwFFHFX+8WzdYtw5yc7/bl1I7n1rN9nDlA8t45Mo+tNOA20SImKCS6cbqS8B5JRw/H+gZfgwHxschJhFJcjk5MGYMpKWVnJwAVqz4fnICCOXVwnY3ouv6Y5WckkzS1KAAzKwL8HYxNajngJnu/lp4ewlwuruvL+F8qkGJVGM5OTBoEGRlHZp4yistDTZtik5cUm5JX4MqTXtgTaHtteF9hzCz4WY2Ny5RiUjCjB0bneQEQZuUJJeqlKDKzN0nuHv/RMchIrE1blx0khMEHSYkuVSlBJUNdCy03SG8T0RqqPLVejz8OFRqKtx0UzQikmiqSglqKnCNBQYBO0tqfxKR6q/stZ4gMXXrZqSmfv9Iaip07w4jR0Y1NImCpElQZvYa8B+gl5mtNbMbzOxGM7sxXORdYAWwHPgLMCJBoYpIkhgxgkMSznf8f48mTYx584x58+C224IOESkpwfNtt8Hs2dCoUfzilrJJql580aZefCJVX07OoYNrR4z4rsYTqRef1c6nfad8FnxdhyZNNOC2CkjugbqxoAQlUrUV14284Lbc7NnB9si7c/nr8ykcyKlDauM8RoyA++6qo1pR1aG5+ESkaimuG3lubrD/4UdDNDpxCR/W/5b0UXV54JIjOa9vW61wW01UuAZlZrWA89z9neiGFD2qQYlUbWlpsGVL8cfrNDxAu5unc0X/jtwx5AiaNqgTv+AkmqJzi8/MBgJXA1cALd09aWthSlAiVVtKCpT462vOv5du5cQereIWk8RExROUmXUlSEo/JpgLr+CEu9y9WZQCjDolKJGqrbQaVKtWzubNup1XDZR/qqNwN+/PCLp23wscDoTCJ/ucYLCsiEhMlNSNPDUVRoxQcqrOSqxBmVmIYCDBAWAGMAmYBmwCprj7sHgEWVGqQYlUbbt2OUcec5Ds1bXwvFr/21+4F5966lULlZosNotgEO0X7l5ChVtEJDpWbM7h56/Nxi75iCPOXUeLliENrq1hSqtBvQxcCjTiu0mslhO0Q73j7hfFOsDKUA1KpOo5mB9iwqcrePpfy6hXO4W7LjiCH/bvqK7j1VvFOkmYWX1gKEEnicF8N3bKgf+6+wlRDDKqlKBEqpaMtTsYNTGTRet3cX7fttx38ZG0blLsXEZSfVS+m7mZpQFXEfTmGwC4u9cq+asSRwlKpGrYeyCPJ6cv5YXPvqVVo3rcf0lfzuvbNtFhSfyUfyYJM3sC+MTd3zKz/kAfd/8j8EczO5wgUYmIVNhny7Zw++QM1mzbx1UDOzH6/N40ra8Bt1K2XnxPufstZvYk8OtkrjEVpRqUSPLasfcAD76ziH9+uZaurRry8LB0BnXTqoE1lObiE5HEc3feyVzPvVMXsH3vQUac3p1fn9WT1DpV5n9fiZOyJChVP0QkKtbv3MfdU+YzY9Em0ts35eWfHk+fdk0SHZYkqbLc4ssCMoF0oBvwVqEi7u6XxTTCStAtPpHkEAo5r/53NY++t5i8UIjfnd2L60/qQu1aSbNmqiRW+XvxhRNUSdSLT0RKtHxTDrdPymDOyu2c3KMVDw1Np1PLBokOS5JLhdqg7otBIBGZ2XnA00At4Hl3f6TI8euAsUB2eNef3f35eMUnIuVzIC/Ec59k8aePllO/bi3G/qAfPziugwbcSpklxYq64bWllgJnA2uBOcBV7r6wUJnrgP7ufnM5zqsalEgCfLNmB6MnZrB4w24u7HcYYy46krTG9RIdliSvCo2DOrW0s7r7pxWNqJCBwHJ3XxF+39eBS4CFJX6ViCSVPfvz+MOHS3nx829p0ziVv1zTn7P7tEl0WFJFlXaLbyYl9+LzMpyjLNoDawptrwWOj1DusnDSXAr8P3dfE6EMZjYcGB6FuESkjD5Zupk7JmWSvWMfVw/qxKjzetM4VQNupeLKmlz2hx+JNA14zd33m9kvgL8BZ0Yq6O4TgAkFt/hEJHa27znAA28vZNLX2XRLa8g/bzyBAV1aJDosqQZK6+OZRXBvMARMBAa7e/PCjyjFkQ10LLTdge86QwDg7lvdvSBJPg8cF6X3FpEKcHfe+iabwU98wtR56/jVmT1499enlCs5bdgAp50WLO1uFjyfdlqwvzhZWdC5c1C+4NG5c7AfICcHxowJluUoWJ5jzJhgv1QtZZnN/GzgJuBCgh52c4HxwN/d/UBUgjCrTXDb7iyCxDQH+JG7LyhU5jB3Xx9+PRQY5e6DSjmvOkmIxED2jn3cNTmTj5ds5qiOzXj0snR6ty3fgNsNG6BLF9gf4d5MvXqwciW0LTJfbFYW9OwJkX6lzeCbb+BHPwrK5eZ+d0wLHCa9ys1mbmbtgBcJeto5cLa7fxS16MyGAE8RJMG/uvuDZnY/MNfdp5rZw8DFQB6wDbjJ3ReXck4lKJEoyg85/zd7FY+9v5iQw63n9uK6E7tQK6X8XcdPOw0+LaGL1amnwieffH9f586wenXxX9OkCRw48P3kVCA1NVjo8L64DZ6RcqjwelCNgWuAG4EjCdqi3gRGunsJFfHEU4ISiZ5lG3czamIGX63ewSk9gwG3HVtUfMBtSkrkmlABMwiFDt1XGWlpsGlT5c4hMVGhmSQmAFcCDYEVwHMEtZttsYgw2pSgRCpvf14+42dm8czHy2lYrzb3XNiHoce0r/SA27J8edFf3comqJQUyM+v3DkkJio81ZED24FZHNrlXHPxiVRjX67azuiJGSzblMMlR7fj7gv70KpRdAbcqgYlhVR4uQ0DWhAMnC1Kf/lFqqGc/Xk8/sES/vaflRzWJJUXrxvAGb1bR/U9Tjml5DaoU045dF+nTpVrg7rppvLHKYlTWg1qTGkncPekbXJUDUqk/D5evIk7J2eyflcu157QhVvP7UWjetFfOk69+KSQyvXiq4qUoETKbmvOfu5/eyFvfbOOHq0b8ehl6RzXObYDbjdsgCuugFmzgqRjFtSc3njj0ORUICsLzjzz+zWpTp3go4+CJJSTA2PHwvjxsHUrtGwZ1JxGjlRySmKVT1Bm9ktgmLufFa2oYkkJSqR07s6Ub7K5f9pCcvbn8cszenDT6d2pVztpV9KR6icqS773AE6vdCgikhTWbNvLnVPm8+nSzRzTqRmPXtaPw9s0TnRYIkB0JnoVkSomP+S89PlKHv9gCSkG9118JFcP6lyhAbcisVLeBLUdKKEPjYgku8UbdjFqYibz1uzg9F5pPDg0nfbN6ic6LJFDqJOESA2xPy+fZz5azriZWTSpX4cxF/Xh4qPaaYVbSQYVWrDwaOBU4C13X1Vof0dgKPCpu38TvRhFJBbmrNzG6IkZZG3ew7Bj2nPXhX1o0bBuosMSKVFp46DeBo4BuhaeudzM6gDLgQx3vyjmUVaQalBS0+3OPchj7y/hldmraN+sPg8NS+e0w9MSHZZIURXqxTcA+KjoshruftDMZgGDoxSciETZjIUbuWvKfDbuzuWnJ3Xld+ccTsMYDLgViZXSPq1NCZa3iMSBZlGNRkQqbfPu/dw3bQFvZ6ynV5vGjL/6WI7pFK21RUXip7RbfMsJklR64aU1wmtDZQA73L1HzKOsIN3ik+oiIyOYYWHXru/2NWkSzMDQr1+w7e68+eVafv/OIvYdyOdXZ/bgF6d1p27t0hbOFkm4Ct3imwb8BvjGzF4AVgGdgRuA5sAr0YxQRA6VkQFHHXXo/l27gv3z5kGz9nu5Y3Imny3fwoAuzXl4WD96tNa8PlK1lVaDakGwxHsXvj9zuQErgQHuvjWG8VWKalBSHTRt+v2a0/c5qQ3y6XzLdGqnpDDq/N78eGAnUjTgVqqWiB/YEuv+4YUJjwcmAOuB/PDzBOCEaCYnMzvPzJaY2XIzGx3heD0zeyN8/Asz6xKt9xZJNllZwfLmZiUlJwAjd28tTu7Rium3nMpPBnWu8skpJwfGjAnWbkpJCZ7HjAn2S82SFAN1zawWsBQ4G1gLzAGucveFhcqMAPq5+41mdiUw1N2vKOW8qkFJlVPSkhKROaEQ1WLAbU4ODBqk5TJqoPLXoADMrH7h2oqZnWBm15pZ1ygGNxBY7u4rwl3aX+fQBRIvAf4Wfv0mcJZVh99IkSLOPLM8yQnAqkVygmCZjKLJCYLtrKzguNQcJSYoMzsD2AxkmdnfzeynwGfAX4FMMxsUpTjaA2sKba8N74tYxt3zgJ1Ayyi9v0jSKGnF2Opu3LjIq+FCsH/8+PjGI4lVWg3qPqABQfXrCuAxYDewJLz/jphGV0FmNtzM5iY6DhEpn62ltGqXdlyql9ISVF/gS4KayjcEXcsvcPc+wBdA/yjFkQ10LLTdIbwvYhkzq00wPivix9XdJ7h7tGITkThpWco9kdKOS/VSWoKqD3zh7tuB/4T3zQ4/fwVEaz3oOUBPM+tqZnWBK4GpRcpMBa4Nv/4BwRRM6v0g1U6nTs73R3WULKUajcMdMSLoEBFJamqwdLvUHKV9tDcCBTNL/ht42d3zw9utCdaHqrRwm9LNwAfAIuAf7r7AzO43s4vDxV4AWoZnt7gFOKQrukhVt3LLHtJ//nV4q2xJavr02MUTbyNHBr31iiapgl58I0cmJi5JjNIG6j4MdHH3q4rsr0swHuoLdx8S2xArTt3MparIyw/x/Gff8uT0pdStlcL1/fry9C3tWL26+N55KSlBcjrzzDgGGgc5OUFvvfHjgzanli2DmtPIkepiXo1F/KBXaBxUuA2oPbArfPsvKSlBSVUwP3snoyZmsGDdLs7p04YHLu1LmybF3OcSqZ4qNBdfRO6eZ2b1gHTg08pEJVJT5R7M58kZS3l+1re0aFiX8T8+lvPTD0t0WCJJo8IzSZjZi8A17l4ruiFFj2pQkqw+z9rC7ZMyWbV1L1f078gdQ46gaYM6iQ5LJFEqtOT7qSUcblupcERqoJ17D/LQu4t4Y+4aOrdswN9/djwn9miV6LBEklJpnSRClNKVSDUokdK5O+/P38A9Uxewbc8Bfn5KN347uCepdZL210cknircBnUQ2Bdhf4Myfr1IjbZxVy53T5nPhws3cmS7Jrx43QD6tm+a6LBEkl5pCWYVkOHuRSdu/V8bVEyiEqkGQiHn9TlrePjdRRzIDzH6/N787OSu1K5VjUbWisRQaQlqJsF6UJFsBWrwtJYixVuxOYfbJ2XyxbfbOKFbSx4elk6XVg0THZZIlZIU60HFitqgJN4O5oeY8OkKnv7XMlJrp3DXBX24vH+HarMchkiMRG8clIgcKmPtDkZNzGTR+l0MSW/LvRcdSWsNuBWpMCUokUraeyCPJ6cv5YXPvqVVo3o895PjOPdIjcIQqSy11kqVlZMDY8ZAWlowL11aGtx+e/AovG/MmKBsLHy2bAvnPvUpf5n1LVcO7MT0W05TchKJErVBSZWUkwODBh26PHhBU0/hH3nBTNizZ0dvstEdew/w+3cW8eaXa+naqiEPD0tnUDctViRSQRHboMpUgzKzZsXs71qJgEQqbOzYQ5MTBImp6P8jublB2bFjK/++7s60eesY/MQnTPk6m1+e0Z33fnOKkpNIDJSpBmVma4Dr3P1fhfb9HHjc3ZN2xKFqUNVXWhps2VL+r9m0qeLvuX7nPu6eMp8ZizbRr0NTHhnWjz7tmlT8hCJSoOLLbYSnPAoBzwBPAn8ELgT2unvSrtCiBFV9paQcWlMqy9fk55derqhQyHn1v6t59L3F5IVC3HpOL647sYsG3IpET6US1CnA80CPQrtnA9e7+9KohBcDSlDVV7xqUMs35TB6YgZzV23n5B6teGhoOp1aNijfSUSkNBVvg3L3WUDBHXwDcoExyZycpHobMeLQZcFLkpoarMpaVgfyQvzpX8sY8vQslm3K4fHLj+KVGwYqOYnEUVlrUNOAIcBeYCLwY4Lk9hd3v7FSAZi1AN4AugArgR9GWqXXzPKBzPDmane/uAznVg2qmoplL76vV29n9MRMlmzczYX9DmPMRUeS1rhedL8BESms0m1Q/wF+4u4rzOx44BWge2WX2zCzx4Bt7v6ImY0Gmrv7qAjlcsrb3qUEVb3l5AQ988aPh61boWVLuOGG4NgLL3y376abYOTI0pPTnv15PP7hEl76fCVtm6TywCV9GdynTey/ERGpVIK6A3jE3UOF9jUExrr7iEpFZbYEON3d15vZYcBMd+8VoZwSlMTMJ0s3c8ekTLJ37OMngzpz23m9aJyqFW5F4qTiCSqWzGyHuzcLvzZge8F2kXJ5wDdAHkGynFLCOYcDw4HjQAlKirdtzwF+//ZCJn2dTfe0hjxyWT8GdGmR6LBEappK1aBWFHPI3b17Gb5+BpGXiL8T+FvhhGRm2929eYRztHf3bDPrBnwEnOXuWaW8r2pQEpG7M3XeOu6btpBd+w4y4vTujDijh1a4FUmMSs1m3qWY/WX6y+/ug4s7ZmYbzeywQrf4InYEdvfs8PMKM5sJHAOUmKBEIsnesY+7Jmfy8ZLNHN2xGY9clk7vthpwK5Jsypqgzij0uhaQDtwLPBGFGKYC1wKPhJ/fKlrAzJoTDAreb2atgJOAx6Lw3lKD5Iec/5u9isfeX0zI4Z4L+3DtiV2olaK1mkSSUYXboMxsLEHnhgGVCsCsJfAPoBPBEvM/dPdtZtYfuNHdf2ZmJwLPEcxmkQI85e4vlOHcusUnACzduJvREzP4avUOTj08jQcv7UvHFhrTJJIkKtUGdU2RXU2A3wEt3T1p740oQcn+vHzGfZzFuJnLaVSvNvdc1IdLj26vFW5Fkkulx0EVLWjAe+5+QeVjiw0lqJrty1XbGT0xg2Wbcrjk6Hbcc2EfWjbSgFuRJFSpThKf8v0EtQf4mui0QYlEVc7+PMa+v5iXZ6/isCapvHjdAM7o3TrRYYlIOSV8HFQsqQZV83y8eBN3Ts5k/a5crj2hC7ee24tG9cr6f5iIJEj5alAR2p0icveXKxqRSLRszdnPfdMWMnXeOnq2bsSbN57IcZ0PGU4nIlVIsTWoYtqdDlHZufhiSTWo6s/dmfx1Ng+8vZCc/Xn88owe3HR6d+rVTtqPpYgcqtxtUEXbnUSSyppte7lzynw+XbqZYzs149HL+tGzTeNEhyUiUaI2KKly8kPOS5+v5PEPlpBiMOr83lx9fGdSNOBWpKoqfy8+M+sHbHb39TEJSaQccnJg1Jhcnn+uFgf2dKFeow7cNAKGpdchRauvi1Q7JdagwosEPuXuvzOzJ4FfuXuV6RKlGlT1sWV7Pn2PyWNTdm0877v2pfIuRigiSalCS75bkS/UPRSJuzkrtzHgitWHJCcIVtPNygoWLhSR6kU3RiRp7c49yF1TMrn82f+w9t/tD0lOBXJzg1V1RaR6Kcvtuu5mdjHQHcDMLqJQTcrdp8YoNqnBZizcyF1T5rNpdy43nNyVMY+VvLrt1q1xCkxE4qa0NqjSxkJ5MrdJqQ2q6tm8ez/3TlvAOxnr6dWmMY/+oB9Hd2xGWhps2VL816WlwaaIK4mJSBVQobn4VqOxUBIH7s6bX67l9+8sYt+BfG4953CGn9qdurWDu9AjRsBjjwW384pKTYWbbopzwCIScxoHJQm3eute7picyacLt9NoSR82zO7Iju1Gy5ZBYho5Mig3aFDQIaJwklIvPpFqoeLLbVRVSlDJLS8/xIv/Xskfpi/B8uqw+x+nsnVdHXJzv/usFk5AEPTWGz8+aHNq2TKoOY0cqeQkUsUpQUnymLNsFz/+9XayPmlLaF9dUutD3kEjL+/QsqmpcNttcN998Y9TROKiQuOgYs7MLjezBWYWCi/zXly588xsiZktN7PR8YxRoif3YD4PTF7CSScay2d0ILSvHmDk7oucnEDdyEVqqmTogTcfGAY8V1wBM6sFPAOcDawF5pjZVHdfGJ8QJRpmr9jK7ZMy+XpyO0I7G+J5Zf//SN3IRWqehCcod18EYFbiJBUDgeXuviJc9nXgEkAJqgrYue8gj7y3mNf+u5pOLRrAoh7kHyxf5b1lyxgFJyJJK+G3+MqoPbCm0Pba8L6IzGy4mc2NeVRSqg8WbODsJz7hjTmrGX5qNz747ans2lG+j526kYvUTHGpQZnZDKBthEN3uvtb0X4/d58ATCjoJCHxt2lXLmOmLuC9+Rs44rAmvHDtANI7NAWC2lBJg24LK+jFV9DVXERqjrgkKHcfXMlTZAMdC213CO+TJOPu/GPuGh58ZxG5eSFuO68XPz+lG3VqfVdrKmnQba1aUK9ecEzdyEVqtqTpZm5mM4Fb3f2QW3NmVhtYCpxFkJjmAD9y9wWlnFPdzONo5ZY93D4pk/+s2MrxXVvw8LB0uqUdmllycjToVkS+J2m7mQ81s7XACcA7ZvZBeH87M3sXwN3zgJuBD4BFwD9KS04SP3n5IcbPzOLcpz5l/rqdPDwsndd+PihicoIg+cyeHYxtSkuDlJTg+bbblJxE5DtJU4OKBdWgYm9+9k5GTcxgwbpdnHtkG+6/pC9tmqQmOiwRqVoqNFmsSET7DuTz1L+W8vysb2nRsC7PXn0s5/U9LNFhiUg1ogQl5fb58i3cPjmTVVv3cuWAjtx+/hE0bVDyek0iIuWlBCVltnPvQR56dxFvzF1Dl5YN+PvPj+fE7q0SHZaIVFNKUFIqd+f9+Ru4Z+oCtu05wI2ndee3g3uSWifyEuwiItGgBCUl2rAzl3vems+HCzfSt30TXrxuAH3bN010WCJSA6gXn0QUCjmvzVnNI+8u5kB+iFvOPpwbTu5K7VoJH5kgItWPevFJ2WRtzuH2SZn899ttnNi9JQ8NTadLq4aJDktEahglKPmfg/khJny6gqf/tYzU2ik8dlk/Lu/fobSZ5kVEYkIJSgDIWLuD297MYPGG3VyQfhhjLu5D68YacCsiiaMEVcPtPZDHk9OX8sJn35LWuB4TfnIc5xwZaeJ5EZH4UoKqwWYt28wdkzNZs20fPzq+E6PP702TVA24FZHkoARVA+3Ye4AH3l7ExK/W0q1VQ94YPojju2nJWhFJLkpQNYi783bGeu6btoAdew9y8xk9uPnMHhpwKyJJSQmqhli/cx93T5nPjEWb6NehKS//9Hj6tGuS6LBERIqlgbrVXCjkvPrFKh59fwl5oRC3ntOL60/qSq0UdR0XkaShgbo1zfJNuxk9MZO5q7ZzSs9WPHhpOp1aNkh0WCIiZaIEVQ0dyAvx7CdZ/Pmj5dSvW4vHLz+Ky45trwG3IlKlKEFVM1+v3s7oiZks2bibi45qxz0X9iGtcb1EhyUiUm4JT1BmdjlwL3AEMNDd5xZTbiWwG8gH8ty9f7xirAr27M/j8Q+X8NLnK2nbJJUXru3PWUe0SXRYIiIVlvAEBcwHhgHPlaHsGe6+JcbxVDkzl2zizsnzyd6xj2tO6MzIc3vRWANuRaSKS3iCcvdFgNpHKmDbngM88PZCJn+dTfe0hrx54wn079Ii0WGJiERFwhNUOTjwYbjr+HPuPqG4gmY2HBget8jizN2ZOm8d901byO7cg/z6rJ788ozu1KutAbciUn3EJUGZ2Qwg0gykd7r7W2U8zcnunm1mrYHpZrbY3T+NVDCcvCYUjIOqTrJ37OOuyZl8vGQzR3dsxqOX9aNX28aJDktEJOrikqDcfXAUzpEdft5kZpOBgUDEBFUd5YecV/6zksc+WALAmIv6cM0JXTTgVkSqrSpxi8/MGgIp7r47/Poc4P4EhxU3SzfuZtTEDL5evYPTDk/jwaF96dBcA25FpHpLeIIys6HAn4A04B0z+8bdzzWzdsDz7j4EaANMDnekqA383d3fT1jQcbI/L59xH2cxbuZyGtWrzVNXHM0lR7dThxIRqRE0F1+S+nLVdkZPzGDZphwuPbodd1/Yh5aNNOBWRKolzcVXFeTsz2Ps+4t5efYq2jWtz4vXD+CMXq0THZaISNwpQSWRjxZv5K7J81m/K5drT+jCref2olE9/YhEpGbSX78ksCVnP/dPW8jUeevo2boRb954Isd1bp7osEREEkoJKoHcnclfZ3P/2wvZsz+P/zf4cG46vTt1a6ckOjQRkYRTgkqQNdv2csfkTGYt28JxnZvzyLB0erbRgFsRkQLqxRdn+SHnpc9X8vgHS0gxGHV+b64+vjMpGnArIjWXevEl2qL1uxg9MYN5a3dyZu/W/P7SvrRrVj/RYYmIJCUlqDjIPZjPMx8vZ/zMLJrWr8MfrzqGi/odpgG3IiIlUIKKsTkrtzF6YgZZm/cw7Nj23H1BH5o3rJvosEREkp4SVIzszj3Io+8v5v9mr6ZD8/q8/NOBnHp4WqLDEhGpMpSgYmD6wo3cPWU+m3bn8rOTu3LLOYfToK4utYhIeeivZhRt3r2fe6ct4J2M9fRu25hnf3IcR3dsluiwRESqJHUzjwJ3559fruXBdxax70A+vz6rB8NP1YBbEZEyUjfzWFi9dS+3T87g38u3MrBLCx4alk6P1o0SHZaISJWnGlQF5eWHePHfK/nD9CXUTklh9Pm9+dHAThpwKyJSfqpBRcuCdTsZPTGTzOydDD6iDb+/tC9tm6YmOiwRkWpFCaoccg/m8/S/ljHh0xU0b1CHZ350LEPS22rArYhIDCQ8QZnZWOAi4ACQBVzv7jsilDsPeBqoRbAU/CPxjHP2iq3cPimTb7fs4Yf9O3DHkCNo1kADbkVEYiXhbVBmdg7wkbvnmdmjAO4+qkiZWsBS4GxgLTAHuMrdF5Zy7kq3Qe3cd5BH3lvMa/9dTacWDXh4WDon9WhV4fOJiMghkrMNyt0/LLQ5G/hBhGIDgeXuvgLAzF4HLgFKTFCV9f78Ddzz1ny25OznF6d247eDD6d+3VqxfEsREQlLeIIq4qfAGxH2twfWFNpeCxxf3EnMbDgwvDKBuDtvfrmWlo3q8cK1A0jv0LQypxMRkXKKS4IysxlA2wiH7nT3t8Jl7gTygFcr+37uPgGYUHCLryLMjD9cfhQN6tWiTi0NuBURibe4JCh3H1zScTO7DrgQOMsjNxhlAx0LbXcI74uppg3qxPotRESkGAmvGoR7590GXOzue4spNgfoaWZdzawucCUwNV4xiohI/CU8QQF/BhoD083sGzN7FsDM2pnZuwDungfcDHwALAL+4e4LEhWwiIjEXsK7mcdSvCaLFRGRSonYzTwZalAiIiKHUIISEZGkpAQlIiJJSQlKRESSUrLNJBETmm1cRCSpubsf8odaNSgREUlK1bqbeTSY2Vx375/oOEqiGKOjKsQIVSNOxRgdNT1G1aBERCQpKUGJiEhSUoIq3YREB1AGijE6qkKMUDXiVIzRUaNjVBuUiIgkJdWgREQkKSlBiYhIUlKCKsLMxprZYjPLMLPJZtasmHLnmdkSM1tuZqPjHOPlZrbAzEJmVmz3TjNbaWaZ4WVM5iZpjIm8ji3MbLqZLQs/Ny+mXH74Gn5jZnFZh6y062Jm9czsjfDxL8ysSzziKmeM15nZ5kLX7mcJiPGvZrbJzOYXc9zM7I/h7yHDzI5NwhhPN7Odha7jPQmIsaOZfWxmC8O/17+JUCb619Ld9Sj0AM4BaodfPwo8GqFMLSAL6AbUBeYBfeIY4xFAL2Am0L+EciuBVgm6jqXGmATX8TFgdPj16Eg/6/CxnDhfu1KvCzACeDb8+krgjSSM8Trgz4n4/BWK4VTgWGB+MceHAO8RLPcwCPgiCWM8HXg7wdfxMODY8OvGwNIIP++oX0vVoIpw9w89WCARYDbB8vJFDQSWu/sKdz8AvA5cEscYF7n7kni9X0WUMcaEXsfwe/0t/PpvwKVxfO+SlOW6FI79TeAsi++cXon+2ZWJu38KbCuhyCXAyx6YDTQzs8PiE12gDDEmnLuvd/evwq93Eywc275IsahfSyWokv2U4D+CotoDawptr+XQH1YycOBDM/vSzIYnOpgIEn0d27j7+vDrDUCbYsqlmtlcM5ttZpfGIa6yXJf/lQn/Q7UTaBmH2A55/7DifnaXhW/3vGlmHeMTWrkk+jNYVieY2Twze8/MjkxkIOHbyccAXxQ5FPVrWSMmiy3KzGYAbSMcutPd3wqXuRPIA16NZ2wFyhJjGZzs7tlm1hqYbmaLw/+tJVOMMVVSjIU33N0LVmCOoHP4OnYDPjKzTHfPinas1dA04DV3329mvyCo8Z2Z4Jiqoq8IPoM5ZjYEmAL0TEQgZtYImAj81t13xfr9amSCcvfBJR03s+uAC4GzPHxztYhsoPB/gx3C+6KmtBjLeI7s8PMmM5tMcFsmagkqCjEm9Dqa2UYzO8zd14dvRWwq5hwF13GFmc0k+O8xlgmqLNeloMxaM6sNNAW2xjCmokqN0d0Lx/M8QZtfson5Z7CyCicCd3/XzMaZWSt33xLPOMysDkFyetXdJ0UoEvVrqVt8RZjZecBtwMXuvreYYnOAnmbW1czqEjRSx6V3V1mZWUMza1zwmqDzR8ReQgmU6Os4Fbg2/Ppa4JBan5k1N7N64detgJOAhTGOqyzXpXDsPwA+KuafqYTFWKT94WKCdotkMxW4JtwDbRCws9Bt36RgZm0L2hfNbCDB3+14/jNC+P1fABa5+xPFFIv+tUxkz5BkfADLCe6jfhN+FPSUage8W6jcEIKeLFkEt7TiGeNQgvu7+4GNwAdFYyToXTUv/FiQjDEmwXVsCfwLWAbMAFqE9/cHng+/PhHIDF/HTOCGOMV2yHUB7if4xwkgFfhn+PP6X6BbPK9dGWN8OPzZmwd8DPROQIyvAeuBg+HP4w3AjcCN4eMGPBP+HjIpoVdsAmO8udB1nA2cmIAYTyZo084o9LdxSKyvpaY6EhGRpKRbfCIikpSUoEREJCkpQYmISFJSghIRkaSkBCUiIklJCUokBsxsppl5eNC3iFSAEpRIBViwlIlHeBwd5ziuMrNPw8sx5JrZKjN73cyOKlTmh2b2tZntMbNdZrbIzJ4udPzeQvE/X2i/WbAUScGx0+P5vYnUyKmORKLobb4/7dHmeL2xmf0R+FV4cxawmGBZhPOBz4B5ZnYcwUDQfQTT1BwE+oTLHLKmD3Clmf3O3XcCg4EeMf0mREqgBCVSOS+4+5TSCoXn8DsNuN7dXwrXRj4GVrl7l/DkxL8nmGXjAjM7BfiEYIaBdHffVuR8x/NdcrrZ3Z8pdKwx381sfhrBnZIX3P03hcq0iBDmdqA5cA3wJ+AmgsRmBDNXiMSVEpRI5dxQ+NaXu/+2gud5BLgAGGJmPwdGhvdfXzQ5hV0cft4AjCt8wIP1enaHNwvmQrvRzLoC/wE+dPcvI5xzFtA3XHYScBHwCjAMJShJALVBiVTOhQS3ygoeFeLu+cBPgD3AcwTLKYxz9w+L+ZLW4edVXjARWrBeVUF70czw8X8QLBlThyDhPATMNbP3w7NTfy+M8Hv3AV4k+Ad2fEW/J5HKUoISqZyh7m4Fj3J8Xa2iOzxYY2oiwS01gOJmjYbvlgYpvLzB6xRZTsXd89396nC5awjaoxw4lyBhFfVXggl+zwa+dPc5pX4nIjGiBCUSH3vCz03Cz32LFgi3K/0YyA3v+mMJ55sWfm5X0JXd3Z8CJhc5Z+/wmlfZ7v6Ku/+IYEZqgMZFT+rBGkNvhjdVe5KEUoISiY+vw8+3mNnjwH2FD4bX7HqFoGZ1CUFb0QVmNjzSydx9Nt+1Pb1oZh+a2Tjg+iJFBwOrzOwjM5tgZm8BRxEkwVnFxHobQQ0qIatJixRQghKJjyeA94FWwBnAk0WOP07Q7vR8uN3peoIedH8ws+6RTujuvwyXm02wbtV1QCPg7wTrMgH8m+C2YVfgaoJefZ8Dl7j7imLOu87dZ7h7bqTjIvGi9aBERCQpqQYlIiJJSQlKRESSkhKUiIgkJSUoERFJSkpQIiKSlJSgREQkKSlBiYhIUlKCEhGRpPT/ATWrUwmaGPuGAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Exclude the anaplerosis reactions\n",
    "reactions_exclude = ['R_OXALODECARB_RXN', 'R_PEPSYNTH_RXN', 'R_PEPDEPHOS_RXN', 'R_PEPCARBOX_RXN', 'R_PEPCARBOX_RXN2', 'R_MALIC_NADP_RXN']\n",
    "\n",
    "# Compare the results and add to list for correlation\n",
    "gsm2_vals = []\n",
    "c13_vals = []\n",
    "c13_vals2 = {}\n",
    "gsm2_vals2 = {}\n",
    "for ind, r in enumerate(c13_model['Reaction ID']):\n",
    "    reacs = r.split(';')\n",
    "    if reacs[0] not in reactions_exclude:\n",
    "        if len(reacs) > 1.0:\n",
    "            sum_flux = 0.0\n",
    "            for reaci in reacs:\n",
    "                reh_name = reac_map[reaci]\n",
    "                sum_flux2 = 0.0\n",
    "                if reh_name != 'missing':\n",
    "                    if ' // ' in reh_name:\n",
    "                        reh_names = reh_name.split(' // ')\n",
    "                        for r in reh_names:\n",
    "                            sum_flux2 += sol_fru2[r]\n",
    "                            if r in change_dir:\n",
    "                                sum_flux2 = sum_flux2*-1.0\n",
    "                                \n",
    "                                \n",
    "    \n",
    "                    else:\n",
    "                        sum_flux2 = sol_fru2[reh_name]\n",
    "                        if reh_name in change_dir:\n",
    "                            sum_flux2 = sum_flux2*1.0\n",
    "                else:\n",
    "                    sum_flux2 = 0.0\n",
    "                            \n",
    "                sum_flux += sum_flux2/2.6 # normalise the flux to 1 mmol/gDCW/h of fructose\n",
    "\n",
    "            c13_vals.append(float(c13_model['13C-MFA'][ind]))\n",
    "            c13_vals2[reaci] = c13_model['13C-MFA'][ind]\n",
    "            gsm2_vals.append(float(sum_flux))\n",
    "            gsm2_vals2[reaci] = float(sum_flux)\n",
    "           \n",
    "        else:\n",
    "            reh_name = reac_map[reacs[0]]\n",
    "            sum_flux2 = 0.0\n",
    "            if reh_name != 'missing':\n",
    "                if ' // ' in reh_name:\n",
    "                    reh_names = reh_name.split(' // ')\n",
    "                    for r in reh_names:\n",
    "                        if r in change_dir:\n",
    "                            sum_flux2 += sol_fru2[r]\n",
    "                            if r in change_dir:\n",
    "                                sum_flux2 = sum_flux2*-1.0\n",
    "                else:\n",
    "                    sum_flux2 = sol_fru2[reh_name]\n",
    "                    if reh_name in change_dir:\n",
    "                            sum_flux2 = sum_flux2*-1.0\n",
    "            else:\n",
    "                sum_flux2 = 0.0\n",
    "                            \n",
    "            c13_vals.append(float(c13_model['13C-MFA'][ind]))\n",
    "            gsm2_vals.append(float(sum_flux2/2.6))\n",
    "            c13_vals2[reacs[0]] = c13_model['13C-MFA'][ind]\n",
    "            gsm2_vals2[reacs[0]] = sum_flux2/2.6\n",
    "                 \n",
    "        \n",
    "        \n",
    "from scipy import stats\n",
    "\n",
    "x = np.linspace(-2, 2,2)\n",
    "y = x\n",
    "plt.plot(x, y)\n",
    "\n",
    "plt.plot(gsm2_vals, c13_vals, 'bo', markersize = 8)\n",
    "axes = plt.gca()\n",
    "\n",
    "plt.rc('axes', linewidth = 2.0)\n",
    "plt.xticks(fontsize=10, weight = 'normal')\n",
    "plt.yticks(fontsize=10, weight = 'normal')\n",
    "\n",
    "axes.spines['right'].set_visible(False)\n",
    "axes.spines['top'].set_visible(False)\n",
    "plt.xlabel('Flux GSM', fontsize = 12, weight = \"bold\")\n",
    "plt.ylabel('Flux 13C-MFA', fontsize = 12, weight = \"bold\")\n",
    "plt.savefig('C13_Correlation.png', format = 'png')\n",
    "plt.tight_layout()\n",
    "slope, intercept, r_value_bio_out, p_value, std_err = stats.linregress(gsm2_vals, c13_vals)\n",
    "print('R correlation coefficient: ', r_value_bio_out, 'p-value: ', p_value)\n",
    "print('R-Squared correlation coefficient: ', r_value_bio_out**2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31d10537",
   "metadata": {},
   "source": [
    "#### Carry out FVA to show variability in model for comparing to 13C data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "38ddf871",
   "metadata": {},
   "outputs": [],
   "source": [
    "C13reacs = [] # only run FVA on the reactions in central carbon metabolism to save time\n",
    "for ind, r in enumerate(c13_vals2.keys()):\n",
    "    if reac_map[r] != 'missing':\n",
    "        reacs = r.split(';')\n",
    "        for reac in reacs:\n",
    "            C13reacs.append(reac_map[reac])\n",
    "            \n",
    "reac_map2 = {}\n",
    "for i, j in reac_map.items():\n",
    "    reac_map2[j] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0c7d3606",
   "metadata": {},
   "outputs": [],
   "source": [
    "res90 = flux_variability_analysis(m2, fraction_of_optimum = 0.9, reaction_list=C13reacs, processes = 6, pfba_factor= 1.2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cf9667ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TKT2 -0.4862838277828041 -0.03965900165647283 -0.008\n",
      "RPE -0.9323499669098846 -0.03910031465722393 -0.008\n",
      "RBPC 0.0 1.3152712470616144 -0.005\n",
      "CS 0.6113109107249172 2.499100720794289 0.393\n",
      "ACONT1 0.6113109107249172 1.5808311286877743 0.393\n",
      "FUM 0.4973604108918595 1.484443933566788 -0.393\n",
      "Number of reactions that agree with the FVA ranges:  25.0\n"
     ]
    }
   ],
   "source": [
    "count = 0.0\n",
    "for ind, i in enumerate(res90['minimum']):\n",
    "    reac_name = reac_map2[res90['minimum'].index[ind]]\n",
    "    if reac_name in c13_vals2:\n",
    "        if res90['minimum'][ind]/2.6 <= c13_vals2[reac_name] <= res90['maximum'][ind]/2.6:\n",
    "            count += 1.0\n",
    "        else:\n",
    "            print(res90['minimum'].index[ind], i/2.6, res90['maximum'][ind]/2.6, c13_vals2[reac_name]) \n",
    "    else:\n",
    "        print(reac_name)\n",
    "            \n",
    "print('Number of reactions that agree with the FVA ranges: ', count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9e64c2",
   "metadata": {},
   "source": [
    "#### 13-C data comparison between models - conclusion\n",
    "\n",
    "Both iCN1361 and RehMBEL1391 show very similar fluxes for central carbon metabolism. When comparing to the 13C data iCN1361 shows a correlation of 0.88 (r-squared), whereas RehMBEL1391 shows a correlation of 0.85 (r-squared). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e925bb73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3906ea00",
   "metadata": {},
   "source": [
    "### Gene knockout phenotype predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e2670a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# load tradis data\n",
    "\n",
    "IPKM_fmm = pd.read_excel('Data/TraDIS_results.xlsx', sheet_name = 'IPKM_FMM')\n",
    "IPKM_sob = pd.read_excel('Data/TraDIS_results.xlsx', sheet_name = 'IPKM_SOB')\n",
    "low_coverage = pd.read_excel('Data/TraDIS_results.xlsx', sheet_name = 'low_coverage')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd90e4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take the logarithm + 1 of the IPKM values for both SOB and FMM media conditions\n",
    "log2_IPKM = {}\n",
    "for ind, gene in enumerate(IPKM_fmm['Unnamed: 0']):\n",
    "    log2_IPKM[gene] = np.log2(IPKM_fmm['IPKM_curated'][ind] + 1)\n",
    "    \n",
    "log2_IPKM_sob = {}\n",
    "for ind, gene in enumerate(IPKM_sob['Unnamed: 0']):\n",
    "    log2_IPKM_sob[gene] = np.log2(IPKM_sob['IPKM_curated'][ind] + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "6635ea0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2/envs/pyTFA_2/lib/python3.7/site-packages/cobra/util/solver.py:554: UserWarning: Solver status is 'infeasible'.\n",
      "  warn(f\"Solver status is '{status}'.\", UserWarning)\n"
     ]
    }
   ],
   "source": [
    "# Add column to excel file with GSM result\n",
    "m2.reactions.Biomass.bounds = (0.0, 1000.0)\n",
    "m2.reactions.FRUpts.bounds = (0.0, 0.0)\n",
    "m2.reactions.FRUpts2.bounds = (0.0, 0.0)\n",
    "m2.reactions.EX_fru_e.bounds = (-2.6, -2.6)\n",
    "\n",
    "# minimal media\n",
    "essential_reh = []\n",
    "non_essential_reh = []\n",
    "\n",
    "sol_wt = pfba(m2)\n",
    "growth_wt = sol_wt['Biomass']\n",
    "for ind, i in enumerate(m2.genes):\n",
    "    #print(ind, i.id)\n",
    "    with m2:\n",
    "        m2.genes.get_by_id(i.id).knock_out()\n",
    "        try:\n",
    "            sol_ko = m2.optimize()\n",
    "            if sol_ko.status != 'infeasible':\n",
    "                if sol_ko['Biomass'] > 0.05:\n",
    "                    #print(sol_ko['R_Biomass'])\n",
    "                    non_essential_reh.append(i.id)\n",
    "\n",
    "                else:\n",
    "                    essential_reh.append(i.id)\n",
    "            \n",
    "            else:\n",
    "                essential_reh.append(i.id)\n",
    "\n",
    "        except:\n",
    "            essential_reh.append(i.id)\n",
    "            \n",
    "            \n",
    "            \n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "2ff5cc6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the false positives, false negatives\n",
    "# How many genes are non-essential but low IPKM\n",
    "false_negatives2 = []\n",
    "false_positives2 = []\n",
    "true_negatives2 = []\n",
    "true_positives2 = []\n",
    "\n",
    "# low coverage genes\n",
    "low_coverage_genes = []\n",
    "for ind, i in enumerate(low_coverage['Low coverage']):\n",
    "    low_coverage_genes.append(i)\n",
    "\n",
    "for gene in non_essential_reh:\n",
    "    if gene not in low_coverage_genes:\n",
    "        if gene in log2_IPKM:\n",
    "            if log2_IPKM[gene] < 4.24:        \n",
    "                #print(gene)\n",
    "                false_negatives2.append(gene)\n",
    "            elif log2_IPKM[gene] > 4.24:\n",
    "                true_negatives2.append(gene)\n",
    "\n",
    "\n",
    "\n",
    "for gene in essential_reh:\n",
    "    if gene != 'Spontaneous':\n",
    "        if gene not in low_coverage_genes:\n",
    "            if gene in log2_IPKM:\n",
    "                if log2_IPKM[gene] > 4.24 and log2_IPKM_sob[gene] > 4.24:        \n",
    "                    #print(gene)\n",
    "                    false_positives2.append(gene)\n",
    "                elif log2_IPKM[gene] < 4.24 or log2_IPKM_sob[gene] < 4.24:\n",
    "                    true_positives2.append(gene)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "b3fa5dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall accuracy RehMBEL:  0.8803212851405623\n",
      "Precision RehMBEL:  0.7407407407407407\n",
      "Recall RehMBEL:  0.5286343612334802\n"
     ]
    }
   ],
   "source": [
    "print('Overall accuracy RehMBEL: ', (len(true_positives2) + len(true_negatives2))/(len(true_positives2) + len(true_negatives2) + len(false_negatives2) + len(false_positives2)))\n",
    "print('Precision RehMBEL: ', len(true_positives2)/(len(true_positives2)+ len(false_positives2)))\n",
    "print('Recall RehMBEL: ', len(true_positives2)/(len(true_positives2) + len(false_negatives2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b447b926",
   "metadata": {},
   "source": [
    "#### Comparison to the results of iCN1361\n",
    "\n",
    "\n",
    "|  | iCN1361 | RehMBEL1391 |\n",
    "|--|---------|-------------|\n",
    "| Overall accuracy | 92% | 88% |\n",
    "| Recall | 62% | 53% |\n",
    "| Precision | 81% | 74% |\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02c6266c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
